### PR TITLE
Refine buffer channel

### PR DESCRIPTION
**Pull request by chengduoZH**
- [x] Send sends only one element, we should wake up only one receiver thread by calling notify_one, instead of notify_all.
- [x] There are two possibilities for the return of `Receive` or `Send`:
    - the `Channel` was closed.
    - the `Receive` or `Send` of this `Channel` was called.
we should separate the two possibilities.

### CODE REVIEW

File: paddle/framework/details/unbuffered_channel.h [hunk 1]
```diff
@@ -66,22 +66,22 @@ void UnBuffered<T>::Send(T* data) {
...
-  if (!closed_) {
-    std::unique_lock<std::mutex> channel_lock(mu_ch_);
-    item = data;
-    channel_lock.unlock();
-    cv_channel_.notify_one();
-    channel_lock.lock();
-    cv_channel_.wait(channel_lock,
-                     [this]() { return item == nullptr || closed_; });
-  }
+  if (closed_) return false;
```
Comment (abhinavarora) [inline paddle/framework/details/unbuffered_channel.h]: We want to make sure that writer_found_ is always false at the end of this function. I would rewrite it as follows:
```
bool ret = false;
if (!closed_) {
    std::unique_lock<std::mutex> channel_lock(mu_ch_);
    item = data;
    channel_lock.unlock();
    cv_channel_.notify_one();
    channel_lock.lock();
    cv_channel_.wait(channel_lock,
                     [this]() { return item == nullptr || closed_; });
    ret = true
  }
  writer_found_ = false;
  return ret;
```

File: paddle/framework/details/unbuffered_channel.h [hunk 2]
```diff
@@ -90,18 +90,18 @@ void UnBuffered<T>::Receive(T* data) {
...
-    // Reader should wait for the writer to first write its data
-    cv_channel_.wait(lock_ch, [this]() { return item != nullptr || closed_; });
-    if (!closed_) {
-      *data = std::move(*item);
-      item = nullptr;
-      lock_ch.unlock();
-    }
-    cv_channel_.notify_one();
-  }
+  if (closed_) return false;
```
Comment (abhinavarora) [inline paddle/framework/details/unbuffered_channel.h]: Similarly I would prefer to rewrite this as follows:

```
bool ret = false;
if (!closed_) {
    std::unique_lock<std::mutex> lock_ch{mu_ch_};
    // Reader should wait for the writer to first write its data
    cv_channel_.wait(lock_ch, [this]() { return item != nullptr || closed_; });
    if (!closed_) {
      *data = std::move(*item);
      item = nullptr;
      lock_ch.unlock();
      ret = true;
    }
    cv_channel_.notify_one();
  }
  reader_found_ = false;
  return ret;
```

File: paddle/framework/details/unbuffered_channel.h (hunk 1 above)
Comment (chengduoZH) [reply paddle/framework/details/unbuffered_channel.h]: Done.

File: paddle/framework/details/unbuffered_channel.h (hunk 2 above)
Comment (chengduoZH) [reply paddle/framework/details/unbuffered_channel.h]: Done.

**Pull request merged**